pkg/runtime/framework/plugins/mpi/mpi.go

```diff
@@ -83,85 +95,113 @@ func (m *MPI) EnforceMLPolicy(info *runtime.Info, trainJob *trainer.TrainJob) er
+						WithSecretName(fmt.Sprintf("%s%s", trainJob.Name, constants.MPISSHAuthSecretSuffix)).
+						WithItems(
+							corev1ac.KeyToPath().
+								WithKey(corev1.SSHAuthPrivateKey).
+								WithPath(constants.MPISSHPrivateKeyFile),
+							corev1ac.KeyToPath().
+								WithKey(constants.MPISSHPublicKey).
+								WithPath(constants.MPISSHPublicKeyFile),
+							corev1ac.KeyToPath().
+								WithKey(constants.MPISSHPublicKey).
+								WithPath(constants.MPISSHAuthorizedKeys),
+						),
+					),
+			}...,
+		)
+		if ps.Name == constants.JobLauncher {
+			apply.UpsertVolumes(
+				&info.TemplateSpec.PodSets[psIdx].Volumes,
+				[]corev1ac.VolumeApplyConfiguration{
+					*corev1ac.Volume().
+						WithName(constants.MPIHostfileVolumeName).
+						WithConfigMap(corev1ac.ConfigMapVolumeSource().
+							WithName(fmt.Sprintf("%s%s", trainJob.Name, constants.MPIHostfileConfigMapSuffix)).
+							WithItems(
+								corev1ac.KeyToPath().
+									WithKey(constants.MPIHostfileName).
+									WithPath(constants.MPIHostfileName).
+									WithMode(0444),
+							),
+						),
+				}...,
+			)
+		}
+		for cIdx, container := range ps.Containers {
+			if container.Name != constants.JobLauncher && container.Name != constants.JobTrainerNode {
+				continue
+			}
+			apply.UpsertVolumeMounts(
+				&info.TemplateSpec.PodSets[psIdx].Containers[cIdx].VolumeMounts,
+				[]corev1ac.VolumeMountApplyConfiguration{
+					*corev1ac.VolumeMount().
+						WithName(constants.MPISSHAuthVolumeName).
+						WithMountPath(*info.RuntimePolicy.MLPolicy.MPI.SSHAuthMountPath),
+				}...,
+			)
+			if ps.Name == constants.JobLauncher && container.Name == constants.ContainerLauncher {
+				apply.UpsertVolumeMounts(
+					&info.TemplateSpec.PodSets[psIdx].Containers[cIdx].VolumeMounts,
+					*corev1ac.VolumeMount().
+						WithName(constants.MPIHostfileVolumeName).
+						WithMountPath(constants.MPIHostfileDir),
+				)
+				switch *info.RuntimePolicy.MLPolicy.MPI.MPIImplementation {
+				case trainer.MPIImplementationOpenMPI:
+					apply.UpsertEnvVars(
+						&info.TemplateSpec.PodSets[psIdx].Containers[cIdx].Env,
+						*corev1ac.EnvVar().
+							WithName(constants.OpenMPIEnvHostFileLocation).
+							WithValue(fmt.Sprintf("%s/%s", constants.MPIHostfileDir, constants.MPIHostfileName)),
+						*corev1ac.EnvVar().
+							WithName(constants.OpenMPIEnvKeepFQDNHostNames).
+							WithValue("true"),
+						*corev1ac.EnvVar().
+							WithName(constants.OpenMPIEnvDefaultSlots).
+							WithValue(strconv.Itoa(int(*info.RuntimePolicy.MLPolicy.MPI.NumProcPerNode))),
+						*corev1ac.EnvVar().
+							WithName(constants.OpenMPIEnvRSHArgs).
+							WithValue("-o ConnectionAttempts=10"),
+					)
+				default:
+					return fmt.Errorf("MPI implementation for %v doesn't supported", info.RuntimePolicy.MLPolicy.MPI.MPIImplementation)
+				}
+			}
+		}
 	}
 
-	// Add container port for the headless service.
-	info.Trainer.ContainerPort = corev1ac.ContainerPort().
```

Comment:
The container port is not required for MPI ?

Reply:
Yes, no need.